doc/design.md

```diff
@@ -0,0 +1,134 @@
+
+Outside of the statsd protocol, agentmon is also capable of handling
+monotonically increasing counters as well. Internally, these are called
+_derived counters_, and they are simply flushed as regular counters. 
+Each time a measurement is observed, the value added to the counter is
+equal to `{observed value} - {previously observed value}`.
+
+In the [Etsy statsd][etsy-statsd] implementation, counters, and timers
+can have an attached sample rate that is typically used to reduce
+observations sent to statsd accepting, possibly, a bit less accuracy
+in reported observations. Statsd timers, however, are not handled in
+agentmon at the time of this writing. 
+
+### Why no timers?
+
+As it turns out, timers just haven't been necessary for the types of
+values we're collecting to date. However, there are also some
+challenges, and incompatibilities with the rest of the Heroku metrics
+infrastructure.
+
+First, the incompatibilities. The Heroku Metrics infrastructure makes
+use of a library called HDR Histogram, which can record the
+distribution of a large number of values at a relatively low cost. The
+cost is typically less than 1KiB in memory, but the serialization format
+is not standardized making it difficult for other agents needing to 
+report to the Heroku Application Metrics endpoints difficult.
+
+An alternative strategy would involve reporting multiple, related
+metrics, namely, sum of squares, sum, min, max, and count for each
+timer metric.  This would allow all of the reporting dynos' values to
+be fairly represented, and we could even derive standard deviation,
+and your typical average latency. However, average latency, and even
+standard deviation of latency isn't all that helpful. What's helpful
+are the tail latencies (p95, p99, p99.9), and often times the median
+latency. These _aren't_ derivable without knowing what the distribution
+looks like, which as we've previously stated, is not representable in
+an easily standardizable way.
+
+## Reporting Metrics to Heroku
+
+When started, the agentmon program expects a URL passed as an argument
+which speaks the protocol outlined below.
+
+While it'd certainly be possible to report metrics to other services,
+the only reporter shipping with agentmon is a simple reporter for
+Heroku's Application Metrics service. This endpoint is POST'ed to over
+HTTPS, with a few optional headers, and a simple JSON body.
+
+The `Content-Type` header should be `application/json`. Optionally,
+the header `Measurements-Time` can be set to an RFC3339 timestamp,
+no more than 4 minutes in the past, representing the time the 
+observations are being reported. The 4 minutes is an implementation
+detail of the Heroku Metrics Aggregation system--metrics timestamped
+older than that will not be aggregated.
+
+Additionally, it's good practice to include the number of measurements
+the payload encodes using the `Measurements-Count` header.
+
+The JSON payload is quite simple:
+
+```json
```

Comment:
since this is invalid JSON, i would kill the `json` bit here

Reply:
Could we make it valid json and still relay the bits?

Reply:
done.